protobuf/lib/src/protobuf/proto3_json.dart

```diff
@@ -88,25 +91,90 @@ Object? _writeToProto3Json(_FieldSet fs, TypeRegistry typeRegistry) {
+            .map((element) => valueToProto3Json(element, fieldInfo.type))
+            .toList();
+      }
+    } else if (fieldInfo.isMapField) {
+      if (context.emitDefaults || (value as Map).isNotEmpty) {
+        skipField = false;
+        jsonValue = (value as PbMap).map((key, entryValue) {
+          var mapEntryInfo = fieldInfo as MapFieldInfo;
+          return MapEntry(convertToMapKey(key, mapEntryInfo.keyFieldType!),
+              valueToProto3Json(entryValue, mapEntryInfo.valueFieldType));
+        });
+      }
+    } else if (_isBytes(fieldInfo.type)) {
+      if (context.emitDefaults || (value as List).isNotEmpty) {
+        skipField = false;
+        if ((value as List).isEmpty) {
+          jsonValue = null;
+        } else {
+          jsonValue = valueToProto3Json(value, fieldInfo.type);
+        }
+      }
+    } else if (fieldInfo.isEnum) {
+      // For enums, the default value is the first value listed in the enum's type definition
+      final defaultEnum = fieldInfo.enumValues!.first;
+      if (context.emitDefaults ||
+          (value as ProtobufEnum).name != defaultEnum.name) {
+        skipField = false;
+        jsonValue = valueToProto3Json(value, fieldInfo.type);
+      }
+    } else if (fieldInfo.isGroupOrMessage) {
+      final originalValue = fs._values[fieldInfo.index!];
+      if (context.emitDefaults || originalValue != null) {
+        skipField = false;
+        if (originalValue == null) {
+          jsonValue = null;
+        } else {
+          jsonValue = valueToProto3Json(value, fieldInfo.type);
+        }
+      }
+    } else if (PbFieldType._baseType(fieldInfo.type) ==
+        PbFieldType._STRING_BIT) {
+      if (context.emitDefaults || (value as String).isNotEmpty) {
+        skipField = false;
+        jsonValue = valueToProto3Json(value, fieldInfo.type);
+      }
+    } else if (PbFieldType._baseType(fieldInfo.type) == PbFieldType._BOOL_BIT) {
+      if (context.emitDefaults || value != false) {
+        skipField = false;
+        jsonValue = valueToProto3Json(value, fieldInfo.type);
+      }
+    } else if (value is Int64) {
+      if (context.emitDefaults || !value.isZero) {
+        skipField = false;
+        jsonValue = valueToProto3Json(value, fieldInfo.type);
+      }
+    } else if (value is int) {
+      if (context.emitDefaults || value != 0) {
+        skipField = false;
+        jsonValue = valueToProto3Json(value, fieldInfo.type);
+      }
+    } else if (value is double) {
+      if (context.emitDefaults || value != 0.0) {
+        skipField = false;
+        jsonValue = valueToProto3Json(value, fieldInfo.type);
+      }
     } else {
-      jsonValue = valueToProto3Json(value, fieldInfo.type);
+      throw JsonSerializationException(
+          'Unexpected field type for proto3 JSON serialization ${fieldInfo.type}');
+    }
+
+    if (!skipField) {
+      result[fieldInfo.name] = jsonValue;
     }
-    result[fieldInfo.name] = jsonValue;
   }
```

Comment:
I discovered an interesting case regarding JSON serialization and proto2. proto2 spec doesn't specify JSON encoding, only proto3 does. However `toProto3Json` can be called on both proto2 and proto3 messages. So if I have a proto2 message like

```proto
syntax = "proto2";

message Test {
  optional int32 a = 1 [default = 123];
}
```

I can convert this to proto3 JSON using `toProto3Json`, but in this PR we assume proto3 defaults -- i.e. for numeric fields the default to be 0. This doesn't hold when the message syntax is proto2. As a result output of this code is somewhat surprising:

```dart
Test test1 = Test.create()..a=0;
print(test1.toProto3Json()); // omits `a`

Test test2 = Test.create()..a=123;
print(test2.toProto3Json()); // generates `a`
```

It's possible to fix this issue and simplify the code at the same time. Here's my suggestion:

```suggestion
  for (var fieldInfo in fs._infosSortedByTag) {
    // Value of the field, or the default for the field if the field is not set
    final dynamic value = fs._getField(fieldInfo.tagNumber);

    // Whether the field should be skipped. In proto3 JSON encoding default
    // values are omitted by default.
    // (https://developers.google.com/protocol-buffers/docs/proto3#json)
    bool skipField = true;

    // Serialized JSON value for the field. Only set and used when `skipField`
    // is false.
    Object? jsonValue;

    if (fieldInfo.isRepeated) {
      if (context.emitDefaults || value != fieldInfo.readonlyDefault) {
        skipField = false;
        jsonValue = (value as PbListBase)
            .map((element) => valueToProto3Json(element, fieldInfo.type))
            .toList();
      }
    } else if (fieldInfo.isMapField) {
      if (context.emitDefaults || value != fieldInfo.readonlyDefault) {
        skipField = false;
        jsonValue = (value as PbMap).map((key, entryValue) {
          var mapEntryInfo = fieldInfo as MapFieldInfo;
          return MapEntry(convertToMapKey(key, mapEntryInfo.keyFieldType),
              valueToProto3Json(entryValue, mapEntryInfo.valueFieldType));
        });
      }
    } else if (_isBytes(fieldInfo.type)) {
      if (context.emitDefaults || !_areListsEqual(value, fieldInfo.readonlyDefault)) {
        skipField = false;
        if ((value as List).isEmpty) {
          jsonValue = null;
        } else {
          jsonValue = valueToProto3Json(value, fieldInfo.type);
        }
      }
    } else if (fieldInfo.isGroupOrMessage) {
      // `_FieldInfo._getField` and `_FieldInfo.readonlyDefault` return empty
      // message instead of `null` when a field with message type is not set,
      // see #309.
      final originalValue = fs._values[fieldInfo.index!];
      if (context.emitDefaults || originalValue != null) {
        skipField = false;
        if (originalValue == null) {
          jsonValue = null;
        } else {
          jsonValue = valueToProto3Json(value, fieldInfo.type);
        }
      }
    } else {
      // enum, string, bytes, numerics, map
      if (context.emitDefaults || value != fieldInfo.readonlyDefault) {
        skipField = false;
        jsonValue = valueToProto3Json(value, fieldInfo.type);
      }
    }

    if (!skipField) {
      result[fieldInfo.name] = jsonValue;
    }
  }
```

I've also added some comments.

You'll notice that two tests will break with this change, but if you look at the test proto you'll see that we actually set the default value as `42` in those tests, so actually we *fix* the test. We should update the `expect`s.